### PR TITLE
Refactor expression parsing thoroughly

### PR DESCRIPTION
**Pull request by Centril**
Based on https://github.com/rust-lang/rust/pull/66994 together with which this has refactored basically the entirety of `expr.rs`.

r? @estebank

### CODE REVIEW

**Comment by Centril** [comment]
Rebased (executed `git rebase --onto stmt-polish{,@{1}}` -- a super handy tool!)

**Comment by Centril** [comment]
https://github.com/rust-lang/rust/pull/66994 has landed so this is ready to go. :)

**Comment by Centril** [comment]
Rebased. :)

File: src/test/ui/issues/issue-13483.stderr [hunk 1]
```diff
@@ -5,10 +5,29 @@ LL |     } else if {
...
+   | |_____^ expected `bool`, found `()`
+
+error[E0308]: mismatched types
+  --> $DIR/issue-13483.rs:11:15
+   |
+LL |       } else if {
+   |  _______________^
+LL | |
+LL | |     }
+   | |_____^ expected `bool`, found `()`
```
Comment (estebank) [inline src/test/ui/issues/issue-13483.stderr]: 🤯❤️

File: src/librustc_parse/parser/expr.rs [hunk 2]
```diff
@@ -1491,72 +1502,71 @@ impl<'a> Parser<'a> {
...
+        let in_span = self.prev_span.between(self.token.span);
+        self.struct_span_err(in_span, "missing `in` in `for` loop")
+            .span_suggestion_short(
+                in_span,
+                "try adding `in` here",
+                " in ".into(),
+                // Has been misleading, at least in the past (closed Issue #48492).
+                Applicability::MaybeIncorrect,
+            )
+            .emit();
```
Comment (estebank) [inline src/librustc_parse/parser/expr.rs]: We could add extra checks here to not suggest `in` when it wouldn't be appropriate and bubble the error up in those cases.

**Review by estebank — APPROVED** [review]

File: src/librustc_parse/parser/expr.rs (hunk 2 above)
Comment (Centril) [reply src/librustc_parse/parser/expr.rs]: What situations do you have in mind for "inappropriate" ?

**Comment by Zoxc** [comment]
Is the `pgo-branch-weights` test unreliable? I see it failed in https://github.com/rust-lang/rust/pull/66942 too in the same worker.

**Comment by Centril** [comment]
> Is the `pgo-branch-weights` test unreliable? I see it failed in #66942 too in the same worker.

@Zoxc Seems like it, cc @michaelwoerister.

**Comment by estebank** [comment]
I've seen that test fail too.

**Pull request merged**